packages/amplify-migration-e2e/src/templategen.ts

```diff
@@ -0,0 +1,175 @@
+import execa from 'execa';
+import path from 'node:path';
+import * as fs from 'fs-extra';
+import { getNpxPath, retry } from '@aws-amplify/amplify-e2e-core';
+import { runGen2SandboxCommand } from '.';
+
+export function runTemplategenCommand(cwd: string, gen1StackName: string, gen2StackName: string) {
+  const parentDir = path.resolve(cwd, '..');
+  const processResult = execa.sync(
+    getNpxPath(),
+    ['--prefix', parentDir, '@aws-amplify/migrate', 'to-gen-2', 'generate-templates', '--from', gen1StackName, '--to', gen2StackName],
+    {
+      cwd,
+      env: { ...process.env, npm_config_user_agent: 'npm' },
+      encoding: 'utf-8',
+    },
+  );
+
+  if (processResult.exitCode !== 0) {
+    throw new Error(`Templategen command exit code: ${processResult.exitCode}, message: ${processResult.stderr}`);
+  }
+}
+
+function uncommentBucketNameLineFromBackendFile(projRoot: string) {
+  const backendFilePath = path.join(projRoot, 'amplify', 'backend.ts');
+  const backendFileContent = fs.readFileSync(backendFilePath, 'utf8');
+  const regex = /^\s*\/\/\s*(s3Bucket\.bucketName)/m;
+  const updatedBackendFileContent = backendFileContent.replace(regex, '$1');
+  fs.writeFileSync(backendFilePath, updatedBackendFileContent);
+}
+
+function setEnvVariable(name: string, value: string) {
+  process.env[name] = value;
+}
+
+function extractContent(readmeContent: string, startRegex: string, endRegex: string) {
+  const pattern = new RegExp(`${startRegex}([\\s\\S]*?)${endRegex}`, 'i');
+  const match = readmeContent.match(pattern);
+
+  if (match && match[1]) {
+    return match[1].trim();
+  }
+  throw new Error('README file parsing failed to get the stack refactor commands');
+}
+
+function extractCommands(readmeContent: string) {
+  const pattern = /```([\s\S]*?)```/g;
+  const matches = readmeContent.matchAll(pattern);
+  const commands = [];
+
+  for (const match of matches) {
+    if (match[1]) {
+      commands.push(match[1].trim());
+    }
+  }
+  if (commands.length === 0) {
+    throw new Error('README file parsing failed to get the stack refactor commands');
+  }
+  return commands;
+}
+
+function getCommandsFromReadme(readmeContent: string) {
```

Comment:
I think the README parsing logic is difficult to follow here. Unless a code reviewer knows a logic beforehand, it is not trivial to know what command 1,2,3 are. Some options here:
1. Black box : Parse the README and get the list of commands, execute them one by one and doesn't care what it is. the test only cares about the end state of the stacks (i.e stack in UPDATE_COMPLETE) and resources are moved/renamed to the right stacks. (Note: You can use stack state to know if update/refactor operation is completed or not, doesn't have to get refactor object or changeset)
2. White-box: parse the README and assume knowledge on the migration logic, like command 1 : upload to s3, command 2: update stack, command 3: refactor, and so on.
```
const stackMigrationHandler = StackMigrationHandler.fromTemplateGenReadme(<readme path>)
stackMigrationHandler.uploadToS3()
stackMigrationHandler.updateStack()
stackMigrationHandler.refactorStacks()
......
```

I would suggest the black box testing for simplicity.

Reply:
That's a great idea. We can do both kinds of testing based on what we are testing. For the happy path, we can do black box testing since we only care about the end state. We already have `getCommandsFromReadme`, we can just execute them one by one.  
For the rollback scenario, we need to extract Rollback specific commands. So that can be its own test.

Reply:
Sounds good, will keep the logic for the `getCommandsFromReadme` and executing commands one by one logic as it is. Should the commands be stored in a variable or accessing them by indices should be okay for this approach?

Reply:
yeah in the black box mode, we don't need the names. Iterating them one by one suffice